src/util/strencodings.h

```diff
@@ -67,6 +68,42 @@ std::vector<Byte> ParseHex(std::string_view hex_str)
 {
     return TryParseHex<Byte>(hex_str).value_or(std::vector<Byte>{});
 }
+
+// Non lookup table version of HexDigit().
+consteval uint8_t ConstevalHexDigit(const char c)
+{
+    if (c >= '0' && c <= '9') return c - '0';
+    // Only lowercase letters are supported, for consistency.
+    if (c >= 'a' && c <= 'f') return c - 'a' + 0xA;
+
+    throw "Called with an invalid hex digit";
+}
+
+/** Converts from hex string literal to byte array at compile time.
+ * @warning Uses a stack based container (array). Care needs to be
+ *          taken to avoid using up too much stack memory especially in
+ *          recursive functions or non-leaf nodes of the call graph.
+ */
```

Comment:
In commit "refactor: Add consteval ArrayFromHex()" (00dc16bc29124dfbaf9d47c5411e4ebcf95e30b5)

This warning seems unnecessarily scary and maybe not true. I don't think the function itself would use any stack space with NRVO, and callers could be calling this to initialize an array on the heap not the stack. Would suggest deleting this comment, because it seems like a warning about std::array in general, and I don't think there is anything that really distinguishes this function from other functions using std::array.

Reply:
This is in response to https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1689162345

To summarize - the concern is not stack memory being used inside `ArrayFromBytes` itself, but rather that the `std::array` returned contains an inner C-array (https://github.com/gcc-mirror/gcc/blob/61715e9340ab8941d40d62158fe437e9dbe3e068/libstdc%2B%2B-v3/include/std/array) that is not allocated from the heap and so will bump the stack pointer in the calling function by a fair bit for long hex strings.

Should clarify the comment a bit on next re-touch.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1716226929

The comment seems not true and not helpful.

- If the concern is _not_ stack space used inside the function, and this warning is only being added because the return type of the function is a variable-sized std::array, then I'm not sure why every function that returns a variable-sized std::array shouldn't have the same warning. This actually seems less deserving of a warning than other functions returning a std::array object, because the size of the returned array is not determined by a number passed by the caller, but by a literal string passed by the caller, and is half the size of that string, so pretty much guaranteed to be small.

- It is also inaccurate to say std::array is a stack based container or that calling this function will use stack space to store the array. This function can be called to initialize not just stack objects, but also global and heap objects due to NRVO, without being allocated on or copied from the stack.

Reply:
New version of the comment in latest push:
```
 * @warning Unlike VecFromHex which returns a vector, the returned array may use
 *          significant stack space if called inside a function.
```

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1717382220

> New version of the comment in latest push:

Thank you! That seems accurate now.